### PR TITLE
[Feature] Adding model_to_pmml

### PR DESCRIPTION
**Pull request by DarkShadeKnigh**
I am aware of the hard coded strings unfortunately that is an unpushed commit on my main desktop I will push it when I am back home otherwise it is ready for testing. Some warning the equal comparison might error due to white space issues.

### CODE REVIEW

File: R/model_to_pmml.R [hunk 1]
```diff
@@ -1,161 +1,530 @@
...
+    # Loop over the list elements from model-steps
+    for (step_name in names(step_list)) {
+      working_step <- step_list[[step_name]]
+      if (length(working_step[["fileType"]]) > 1) {
+        for (sub_step_index in 1:length(working_step[["fileType"]])) {
+          tmp_file_type <- working_step[["fileType"]][[sub_step_index]]
+          if (is.na(tmp_file_type)) {
+            tmp_file_type <- step_name
+          }
+          tmp_file <- working_step[["file"]][[sub_step_index]]
```
Comment (yulric) [inline R/model_to_pmml.R:167]: Rename to `current_file_path`

File: R/model_to_pmml.R [hunk 2]
```diff
@@ -1,161 +1,530 @@
...
-    # Add the Apply node and two constant Constant nodes as children to the DerivedField node
-    # The 2 constant Constant nodes have dataType of float and values of 0 and 1 
-    # Add the DerivedField to the return PMML
-    # At the end of the loop return the PMML
+    for (vector_index in 1:length(origVariable)) {
+      # For each iteration of the loop create a DerivedField node
+      # The name attribute is set using dummy_variable value
+      # the optype attribute is set to categorical
+      # The dataType attribute is set to integer
+      tmp_main_node <-
```
Comment (yulric) [inline R/model_to_pmml.R:261]: Rename to `current_derived_field_node`

File: R/model_to_pmml.R [hunk 3]
```diff
@@ -1,161 +1,530 @@
...
-  # Add the DerivedField as a child node to TransformationDictionary
-  
+  for (vector_index in 1:length(variable)) {
+    variable_list <- strsplit(rcsVariables[[vector_index]], ";")[[1]]
+    # Create the first rcs DerivedField node
+    # Create a temporary Constant Array Node with n set to 5, type set to float, and values from knots
+    tmp_knots <- strsplit(knots[[vector_index]], ";")[[1]]
+    tmp_knots <- paste(tmp_knots, collapse = " ")
+    tmp_array_node <-
+      XML::xmlNode("Array", attrs = c(n = 5, type = "float"), tmp_knots)
```
Comment (yulric) [inline R/model_to_pmml.R:394]: I made a mistake when creating the test for this. The `n` attribute should be set to the number of knots.

File: R/model_to_pmml.R [hunk 4]
```diff
@@ -1,161 +1,530 @@
...
+    for (vector_index in 1:length(interactionVariable)) {
+      variable_list <-
+        strsplit(interactingVariables[[vector_index]], ";")[[1]]
+      # Create DerivedField node with name attribute set by interaction_variable column
+      # optype attribute is determined based on interaction_variable_type cont= continuous, cat = categorical
+      # dataType attribute is determined based on interaction_variable_type cont= float, cat = string
+      optype <- ""
+      data_type <- ""
+      switch (interactionVariableType[[vector_index]],
+              "cont" = {
```
Comment (yulric) [inline R/model_to_pmml.R:493]: This should be pulled out into its own function and called in all the create node functions

File: R/model_to_pmml.R [hunk 5]
```diff
@@ -172,50 +541,207 @@ create_interaction_nodes <-
...
+    
     # Create ParameterList, FactorList, CovariateList, ParamMatrix child nodes for GeneralRegressionModel
-    # Repeat above step with name set to time and usageType to active
+    tmp_param_list_node <- XML::xmlNode("ParameterList")
+    tmp_factor_list_node <- XML::xmlNode("FactorList")
+    tmp_covariate_list_node <- XML::xmlNode("CovariateList")
+    tmp_param_matrix_node <- XML::xmlNode("ParamMatrix")
+    tmp_pp_matrix_node <- XML::xmlNode("PPMatrix")
     # Create a ParameterList child node Paramater with name set to p0 and label to Intercept
+    tmp_param_list_node <-
```
Comment (yulric) [inline R/model_to_pmml.R:563]: Move this and the next line to below the line where you create the `ParameterList` node

File: R/model_to_pmml.R [hunk 6]
```diff
@@ -172,50 +541,207 @@ create_interaction_nodes <-
 #'
 #' @return returns PMML with attached nodes correlating to the beta_coefficient
 create_beta_coefficient_nodes <-
-  function(variable, coefficient, type, start_variables, PMML) {
+  function(variable,
+           coefficient,
+           type,
+           start_variables,
```
Comment (yulric) [inline R/model_to_pmml.R:547]: Rename to `mining_fields`

File: R/strings.R [hunk 7]
```diff
@@ -18,11 +18,11 @@ pkg.globals$argument.VariableStartLow <- "low"
 pkg.globals$argument.CatStartValue <- "value"
 pkg.globals$argument.CatStartLabel <- "valueLabelStart"
 pkg.globals$argument.VariableStartLabel <- "label"
-pkg.globals$argument.From <- "recFrom"
+pkg.globals$argument.From <- "recStart"
```
Comment (yulric) [inline R/strings.R:21]: Should be called recStart

File: R/strings.R [hunk 8]
```diff
@@ -18,11 +18,11 @@ pkg.globals$argument.VariableStartLow <- "low"
...
 pkg.globals$argument.Interval <- "interval"
-pkg.globals$argument.CatValue <- "recTo"
+pkg.globals$argument.CatValue <- "recEnd"
```
Comment (yulric) [inline R/strings.R:23]: Should be called recEnd

**Review by yulric — CHANGES_REQUESTED** [review]
1. For all the functions to create the PMML nodes, can you rename the function arguments to use snake case
2. Before you let me know to review it again, please fix all the hardcoded strings
3. Can you rename all variables that have the word `tmp` in them? The variables you're calling tmp are not actually temporary which is giving them a false meaning
4. Can you add a `expect_xml_equal` helper function for the tests and use it? There's a lot of duplicate code for checking if 2 XML string are equal that should go into this function

**Comment by yulric** [comment]
@DarkShadeKnigh I also pushed an update for the dummy step conversion. There was an error in the expected PMML and the spec. Can you update the code before telling me to review it again?

File: R/strings.R [hunk 9]
```diff
@@ -105,9 +105,44 @@ pkg.globals$ModuleTypes.Function <- "function"
...
-pkg.globals$WorkingData.ModuleOperations <- "ModuleOperations"
+
 
 # FunctionList contents
 pkg.globals$FunctionList.Arguments <- "arguments"
 pkg.globals$FunctionList.Parameter <- "parameter"
 pkg.globals$FunctionList.VariableArguments <- "variablesArguments"
+
+# PMML Export
+pkg.globals$PMML.variables <- "variables"
```
Comment (yulric) [inline R/strings.R:116]: The next 3 lines should be scoped to ModelExportCSV, so `pkg.globals$ModelExportCSV`

File: R/strings.R [hunk 10]
```diff
@@ -105,9 +105,44 @@ pkg.globals$ModuleTypes.Function <- "function"
...
+pkg.globals$PMML.variable_details <- "variable-details"
+pkg.globals$PMML.model_steps <- "model-steps"
+
+pkg.globals$PMML.filePath <- "filePath"
```
Comment (yulric) [inline R/strings.R:120]: The next 3 lines should be scoped to `ModelStepsCSV`

File: R/model_to_pmml.R [hunk 11]
```diff
@@ -1,38 +1,207 @@
...
+        list(fileType = current_rows[[pkg.globals$PMML.fileType]], filePath = current_rows[[pkg.globals$PMML.filePath]])
+    }
+    
+    # Loop over above list and read in each file specified saving it in a list containing the step name and fileType as well as the content of the file
+    if (length(step_list) >= 1) {
+      for (single_step in 1:length(step_list)) {
+        # Check for sub steps
+        if (length(step_list[[single_step]][[1]]) > 1) {
+          for (sub_step in 1:length(step_list[[single_step]][[1]])) {
+            current_file_path <-
```
Comment (yulric) [inline R/model_to_pmml.R:69]: Can we move lines 69 to 79 and lines 81 to 87 into a function? They're doing the same thing.

File: R/model_to_pmml.R [hunk 12]
```diff
@@ -1,38 +1,207 @@
...
+          )
+      }
+    }
+    
+    # Loop over the list elements from model-steps
+    for (step_name in names(step_list)) {
+      working_step <- step_list[[step_name]]
+      if (length(working_step[[pkg.globals$PMML.fileType]]) > 1) {
+        for (sub_step_index in 1:length(working_step[[pkg.globals$PMML.fileType]])) {
+          current_file_type <-
```
Comment (yulric) [inline R/model_to_pmml.R:171]: Can we move lines 171 to 185 and lines 187 to 201 into a function? They're both doing the same thing.

**Review by yulric — CHANGES_REQUESTED** [review]
Let's move all those hardcoded strings to the strings.R file. A lot of them are repeated.

**Comment by yulric** [comment]
1. It looks like you're missing the function that creates a targets list from modules map CSV or a module CSV file
2. The column header for the CSV file should use the same naming convention as recodeflow (snake case)
3. The spec should be updated to split the `step_arguments` column into the columns `stepArgumentName` and `stepArgumentValue`
4. Can you update the spec with the description of the modules map file
5. I think your tests are too complicated for testing what you want. Don't use the depression imputation module but create simpler module to specifically test what you want.
6. Can you also add the flow diagram for the verify_targets function?
